### ISSUE TITLE
[1.4.2.x] binaries in dev bundle (i.e. `meteor npm i -g`) missing from PATH

### ISSUE THREAD

**Issue by tcastelli** [meteor/meteor#8277]
This seems to be still happening on Windows as checked here (https://forums.meteor.com/t/yarn-and-meteor/30254/15), after doing `meteor npm install -g yarn|jsdoc|whatever` , trying to run meteor "command" fails, since the installation is not performed in the correct dir (files get copied to the root of dev_bundle instead of the bin folder). 

EDIT: I used to think some packages were working, but seems that I installed them with a previous version of Meteor (which was working and placing bins on the correct folder) but with 1.4.2.3 this seems no longer the case and all get installed under the wrong folder.

EDIT2: Tried also with latest beta and problem still persists

**Comment by abernix** [comment]
Thanks for opening this.  I ran into this in another issue last week, but I can't find the issue at the moment.  I looked into it briefly and it could actually be a Windows Node issue.  While Meteor does some [creative stuff](https://github.com/meteor/meteor/blob/devel/tools/cli/dev-bundle-bin-helpers.js#L110-L112) on Windows in regards to `node-gyp` compilation, it shouldn't really be causing this so I was inclined to believe that this could be a Node difference between platforms.

Marking as confirmed as the reproduction above was enough to replicate, but this will have to wait until I hop on Windows again.

**Comment by tcastelli** [comment]
Maybe uptating node to latest 4.7.x would help then?

Regarding the topic change, I think this problem is not for 1.4.x in general, since at some point it was working (and I believe it was 1.4.2 maybe due a previous version of Node doing installs on the proper place). In any case one solution would be to add to PATH for execution `mt-os.windows.x86_32\dev_bundle\` folder directly in addition to `mt-os.windows.x86_32\dev_bundle\bin`, but sounds a little bit dirty since we will end with two node_modules folders and two places for binaries.

**Comment by abernix** [comment]
Some notes, I just found showing some differences between the way `meteor npm` and `npm` worked (with matched versions) on Mac vs PC.

These are in a standardized pseudo form – that is to say, `<METEOR-TOOL>` was the actual full path to my Meteor tool (So, `C:\Users\IEUser\AppData\Local\.meteor\packages\meteor-tool\1.4.2_3\mt-os.windows.x86_32\dev_bundle\bin`, on Windows for Meteor 1.4.2.3)

#### Windows
* `meteor npm`:
  * `<METEOR-TOOL>/dev_bundle/bin`
  * `<METEOR-TOOL>/dev_bundle/lib/node_modules/.bin`
* `npm`:
  * `<SYSTEM-NPM>/npm/node-gyp-bin`
  * `<APP-DIR>/node_modules/.bin`

#### Mac
* `meteor npm`:
  * `<METEOR-TOOL>/dev_bundle/lib/node_modules/npm/bin/node-gyp-bin`
  * `<APP-DIR>/node_modules/.bin`
  * `<METEOR-TOOL>/dev_bundle/bin`
  * `<METEOR-TOOL>/dev_bundle/lib/node_modules/.bin`

* `npm`:
  * `<SYSTEM-NPM>/npm/node-gyp-bin`
  * `<APP-DIR>/node_modules/.bin`

@tcastelli In regards to this being a specific regression in a specific version of Meteor (or Node), if you have a moment to check specific Meteor versions (using `--release w.x.y.z`), that'd be awesome.

**Comment by tcastelli** [comment]
I was checking the history of the file to see if there was a significant change between Meteor 1.4.x versions but I think not, so gonna try with different Meteor versions to see if Node was the culprit.

**Comment by abernix** [comment]
@tcastelli :

> I think I have found another different bug since typing meteor --release 1.4.2 npm install -g jsoc returns 'npm' is not a Meteor command. See 'meteor --help' is that supposed to work

Eeps, yeah, that actually won't work (despite my previous suggestion) because when running built-in commands Meteor skips the overhead of running [`main.js`](https://github.com/meteor/meteor/blob/devel/tools/cli/main.js#L521) which handles the [springboarding](https://github.com/meteor/meteor/blob/devel/tools/cli/main.js#L422) to a specific Meteor release.  You should be able to test the different versions by using:

    %LOCALAPPDATA%\.meteor\packages\meteor-tool\1.4.2_1\os.windows.x86_32\meteor.bat npm

...where `1.4.2_1` is one of the already-installed versions (you'll note there's an interesting naming convention there with the `_`) you have in your `%LOCALAPPDATA%\.meteor\packages\meteor-tool` directory.  You can download other versions first by running `meteor --release <version> --help`. 

> And btw, do you mean that by running meteor npm on Mac for example the bin file gets copied to 4 different places?

The bullet-points I listed above are directories in the search `PATH`, not locations where the binaries are copied to.

**Comment by tcastelli** [comment]
I see, I will try with previous verions in that way (from the help menu that skipping is not really clear, so maybe help menu should be updated).

And in terms of the PATH, that looks strange since I would expect meteor npm to add the same folders on the different operative systems 💩

**Comment by abernix** [comment]
> (from the help menu that skipping is not really clear, so maybe help menu should be updated).

Not immediately clear to me what you mean by this.  Can you clarify?

> I would expect meteor npm to add the same folders on the different operative systems

Yes, me too – that is why I laid out the differences exhibited above.  I didn't look into why they were different yet, only that they were.  It's possible that the difference is due to an issue in either NPM or Meteor.  Your help in investigating is appreciated!

**Comment by tcastelli** [comment]
I mean, that meteor  with `meteor --help`, maybe another line should clarify that --release doesnt work with non built-in commands (if that's how it's going to be tackled) so maybe adding something like this:
`Usage: meteor [--release <release>] [--help] <built-in command> [args]
       meteor <npm command> [args]
       meteor help <command>
       meteor [--version] [--arch]
`

**Comment by tcastelli** [comment]
I've got some time today and could verify if this was indeed working on previous versions. I have some good and bad news :)

**The bad:**
`C:\Users\Dell\AppData\Local\.meteor\packages\meteor-tool\1.4.2\mt-os.windows.x86_32\meteor.bat npm -g install jsdoc`  installs jsdoc binaries on my latest meteor version directory and not on its own, in this case it targeted 1.5.0-beta.4 folder.

**The good:**
The binaries are placed in the proper folder `C:\Users\Dell\AppData\Local\.meteor\packages\meteor-tool\1.5.0-beta.4\mt-os.windows.x86_32\dev_bundle\bin` and the command can be run from console. 😃 

BTW. I can also confirm that meteor npm install -g with 1.5 releases are also not working correctly in Windows since binaries end in the root of the dev_bundle. Seems like some root path must be set for npm in order to place the binaries properly with the latest versions. (Not sure when they changed that exactly)

**Comment by abernix** [comment]
@tcastelli 

> I mean that running meteor --help, maybe should give another line to clarify that

This is a great suggestion!  Do you want to submit a pull-request that makes that change to [`/tools/cli/help.txt`](https://github.com/meteor/meteor/blob/devel/tools/cli/help.txt#L1-L8)?

> I've got some time today and could verify if this was indeed working on previous versions.

In your "The bad" statement, you show yourself using 1.4.2.  Did you test other versions, or just 1.4.2?  You should be able to check the specific version of NPM bundled with:

    C:\Users\Dell\AppData\Local\.meteor\packages\meteor-tool\1.4.2\mt-os.windows.x86_32\meteor.bat npm --version

It would be interesting to correlate that with NPM or Node's changelog.

**Comment by tcastelli** [comment]
Ok, I will check in the history if npm version was changed in 1.4.2.1 and 1.4.2.2 and if so, will try to run again install -g to see where it stoped working (1.4.2.3 fails for sure).

And regarding submiting a PR, Do you think it is reasonable that the help text shows
```
List of detected external commands:
   jsdoc                XXX
   yarn              XXX
  ...
```
so before printing help we use `meteor npm list -g -depth 0` to parse available commands?
(I know list -g would not work on Windows right now, but should be the command to use once this is solved if I'm not wrong)

**Comment by abernix** [comment]
Reasonable idea!  This is already too far off-topic for this issue, but currently there's no way to just "get" the external commands – they are simply checked to see [if they exist at run-time](https://github.com/meteor/meteor/blob/release-1.4.3/tools/cli/dev-bundle-bin-helpers.js#L13-L42).  We could then offer instructions in that section: ``No external commands installed, for example, to install `yarn`, you can run `meteor npm install --global yarn` `` If you wanted to engineer something that would offer that (by reading `devBundleDir`), it would be great, but maybe open another issue first? :smile:

Still off topic, but I think a PR that adds the changes you suggested above, but instead of `meteor <external command> [args]` just explicitly puts `meteor node [args]` and `meteor npm [args]` would be good!

**Comment by tcastelli** [comment]
I have sent an initial PR #8355 to include the basic changes proposed previously. Any further enhancements could be added later, once this issue gets solved (hopefully).

Besides that, I have verified that things get broken at 1.4.2.1 and that there must be something weird affecting npm starting that meteor version, since the npm version is shipped is `3.10.9` at both 1.4.2 and 1.4.2.1

@abernix  any suggestion on how to track which difference on both version could be "activating this different behaviour"?

**Comment by hwillson** [comment]
There hasn't been any activity on this issue in a while, and there have been several changes in this area since this issue was reported. I'll close this issue for now, but if anyone is still seeing this problem using Meteor >= 1.5.2, please post back and I'll re-open. Thanks!

**Comment by nathantreid** [comment]
I'm on Meteor 1.5.2.2, and I'm still having this issue. Hadn't tried it in a while, but one of my coworkers brought up using Yarn so I gave it another shot. Here's the output:
``` 
C:\projects\go2\swmid-418710\site>meteor npm i -g yarn
H:\.meteor\packages\meteor-tool\1.5.2_2\mt-os.windows.x86_32\dev_bundle\yarn -> H:\.meteor\packages\meteor-tool\1.5.2_2\mt-os.windows.x86_32\dev_bundle\node_modules\yarn\bin\yarn.js
H:\.meteor\packages\meteor-tool\1.5.2_2\mt-os.windows.x86_32\dev_bundle\yarnpkg -> H:\.meteor\packages\meteor-tool\1.5.2_2\mt-os.windows.x86_32\dev_bundle\node_modules\yarn\bin\yarn.js
H:\.meteor\packages\meteor-tool\1.5.2_2\mt-os.windows.x86_32\dev_bundle
`-- yarn@1.2.1



   ╭─────────────────────────────────────╮
   │                                     │
   │   Update available 4.6.1 → 5.5.1    │
   │     Run npm i -g npm to update      │
   │                                     │
   ╰─────────────────────────────────────╯


C:\projects\go2\swmid-418710\site>meteor yarn
'yarn' is not a Meteor command. See 'meteor --help'.
```

**Comment by Rafael-Menis** [comment]
I'm still running into this issue on meteor 1.7.0.5, on Windows 10 64 bit. I've tried manually downgrading the system node version to 8.11.4 and reinstalling yarn without dependencies (as Mcissel reccomended back in 2017- http://mcissel.net/using-yarn-with-meteor/, since meteor node -v returns 8.11.4), but I still get "'yarn' is not a Meteor command. See 'meteor --help'.". Even after editing my PATH and manually putting yarn into what should be the appropriate location (.meteor/local/dev_bundle/lib/node_modules), I still get that same error. Is there any way around this?
